File: python_mvp/plant.py
import numpy as np
import math

# TODO: You could also model accelerometer accuracy by rounding up/down to certain +-g numbers
# This is just 1d because i think its okay to assume that accel_y will always be in the direction of -g
class AccelerometerPlant:
    def __init__(self, noise_std_dev, iteration_latency):
        self.noise_std_dev = noise_std_dev
        self.iteration_latency = iteration_latency
        self.actual_acceleration = 0
    
    def apply_force(self, net_force, mass):
        self.actual_acceleration = net_force / mass
    
    def get_acceleration(self):
        return np.random.normal(0.0, self.noise_std_dev, size=1)[0] + self.actual_acceleration

# This is just going to give back altitude instead of pressure which needs to be converted because I'm assuming the sensor is doing this for us, if not its not a crazy thing to be adding
class BarometerPlant:
    def __init__(self, noise_std_dev, iteration_latency):
        self.noise_std_dev = noise_std_dev
        self.iteration_latency = iteration_latency
        self.actual_altitude = 0
    
    def apply_accel(self, accel, velocity, dt):
        self.actual_altitude += velocity * dt + 0.5 * accel * dt**2
    
    def get_altitude(self):
        return np.random.normal(0.0, self.noise_std_dev, size=1)[0] + self.actual_altitude

class RocketPlant:
    def __init__(self, initial_mass, cD, A, rho, noise_std_dev, iteration_latency, initial_velocity):
        self.barometer_plant = BarometerPlant(noise_std_dev, iteration_latency)
        self.accelerometer_plant = AccelerometerPlant(noise_std_dev, iteration_latency)
        
        self.initial_mass = initial_mass
        self.propellant_mass = 8.0
        self.cD = cD
        self.A = A
        self.rho = rho
        
        self.total_impulse = 5000
        self.burn_time = 8.0
        self.average_thrust = self.total_impulse / self.burn_time
        self.max_thrust = self.average_thrust * 1.5
        
        # State variables
        self.noise_std_dev = noise_std_dev
        self.iteration_latency = iteration_latency
        self.initial_velocity = initial_velocity
        self.a = 0
        self.v = initial_velocity
        self.y = 0
        self.t = 0
   
    # These mass and thrust functions + constants are taken from claude so I can see if my idea works at all, these will need to be changed
    def __mass(self):
        if self.t <= self.burn_time:
            consumed_propellant = (self.t / self.burn_time) * self.propellant_mass
            return self.initial_mass - consumed_propellant
        else:
            return self.initial_mass - self.propellant_mass
    
    def __thrust(self):
        if self.t > self.burn_time:
            return 0.0
        
        burn_fraction = self.t / self.burn_time
        
        if burn_fraction < 0.1:
            return self.average_thrust * 1.5
        elif burn_fraction < 0.8:
            return self.average_thrust
        else:
            tail_fraction = (1.0 - burn_fraction) / 0.2
            return self.average_thrust * tail_fraction
    
    def update(self, dt):
        self.t += dt
        
        F_thrust = self.__thrust()
        
        drag_direction = -1 if self.v >= 0 else 1
        F_drag = drag_direction * 0.5 * self.cD * self.A * self.rho * self.v**2 
        
        F_g = self.__mass() * 9.81
        F_net = F_thrust - F_drag - F_g
        
        self.a = F_net / self.__mass()
        self.v += self.a * dt
        self.y += self.v * dt + 0.5 * self.a * dt**2
        
        self.barometer_plant.apply_accel(self.a, self.v, dt)
        self.accelerometer_plant.apply_force(F_net, self.__mass())

File: hardware_sim/src/plant.cc
#include "plant.h"
#include <Arduino.h>

RocketDynamics::RocketDynamics(DynamicsParams params) : params_(params) {}

const Matrix<3, 1> RocketDynamics::Dynamics(const Matrix<3, 1> &X, float32_t t,
                                            float32_t dt) const {
  AccelFunction accel = [&](float32_t t, float32_t v) -> float32_t {
    const float32_t rocket_mass = mass(t);
    const float32_t F_thrust = thrust(t);

    const int sign = v < 0 ? -1 : 1;
    const float32_t F_drag =
        -0.5 * params_.cD * params_.A * params_.rho * v * abs(v);
    const float32_t F_weight = rocket_mass * 9.81;
    const float32_t F_net = F_thrust - F_drag - F_weight;

    return F_net / rocket_mass;
  };

  const Matrix<2, 1> X_mini{X.at(States::Y, 0), X.at(States::V, 0)};

  const Matrix<2, 1> X_mini_out = rk4_step(t, dt, X_mini, accel);

  const Matrix<3, 1> X_t{X_mini_out.at(States::Y, 0),
                         X_mini_out.at(States::V, 0),
                         accel(t + dt, X_mini_out.at(States::V, 0))};

  return X_t;
}

const Matrix<2, 1> RocketDynamics::rk4_step(float32_t t, float32_t dt,
                                            const Matrix<2, 1> &X,
                                            AccelFunction accel) const {
  const float32_t v0 = X.at(States::V, 0);
  const float32_t y0 = X.at(States::Y, 0);

  const float32_t v1 = v0;
  const float32_t a1 = accel(t, v1);

  const float32_t v2 = v0 + (a1 * dt / 2.0);
  const float32_t a2 = accel(t + (dt / 2.0), v2);

  const float32_t v3 = v0 + (a2 * dt / 2.0);
  const float32_t a3 = accel(t + (dt / 2.0), v3);

  const float32_t v4 = v0 + (a3 * dt);
  const float32_t a4 = accel(t + dt, v4);

  const Matrix<2, 1> output({y0 + (dt / 6.0) * (v1 + 2 * v2 + 2 * v3 + v4),
                             v0 + (dt / 6.0) * (a1 + 2 * a2 + 2 * a3 + a4)});

  return output;
}

float32_t RocketDynamics::mass(float32_t t) const {
  if (t <= params_.burn_time) {
    float32_t propellant_consumed =
        (t / params_.burn_time) * params_.propellant_mass;

    return params_.initial_mass - propellant_consumed;
  } else {
    return params_.initial_mass - params_.propellant_mass;
  }
}

float32_t RocketDynamics::thrust(float32_t t) const {
  if (t > params_.burn_time) {
    return 0.0;
  }

  const float32_t burn_fraction = t / params_.burn_time;

  const float32_t average_thrust = params_.total_impulse / params_.burn_time;

  if (burn_fraction < 0.1) {
    return average_thrust * 1.5;
  } else if (burn_fraction < 0.8) {
    return average_thrust;
  } else {
    const float32_t tail_fraction = (1.0 - burn_fraction) / 0.2;
    return average_thrust * tail_fraction;
  }
}

RocketPlant::RocketPlant(RocketDynamics::DynamicsParams params)
    : dynamics_(params), a_(0), v_(0), y_(0), t_(0) {}

void RocketPlant::Update(float32_t dt) {
  const Matrix<3, 1> X{y_, v_, a_};

  const Matrix<3, 1> X_t = dynamics_.Dynamics(X, t_, dt);

  t_ += dt;

  a_ = X_t.at(RocketDynamics::States::A, 0);
  v_ = X_t.at(RocketDynamics::States::V, 0);
  y_ = X_t.at(RocketDynamics::States::Y, 0);
}

float32_t RocketPlant::accel() const { return a_ + random(-3, 3); }

float32_t RocketPlant::position() const { return y_ + random(-10, 10); }
